external authz: Interpret error grpc codes of the external authz server as failure to fix failure_mode_allowed feature

konstantin-baidin-y42 opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message:  Interpret error grpc codes of the external authz server as failure to fix failure_mode_allowed feature
Additional Description: Fixes issue https://github.com/envoyproxy/envoy/issues/34705. Mark authz response as denied only for grpc codes `PermissionDenied` and `Unauthenticated`. Other codes are considered as failure and authz response status is set to error. This new behaviour is aligned with the HTTP external authz checks implementation, where error HTTP codes are considered as failure too. It shouldn't change anything when `failure_mode_allowed=false` as any non-OK grpc codes will lead to 403 for a client. But it fixes behaviour of `failure_mode_allowed=true`. 
Risk Level: Low
Testing: unit test
Docs Changes: ---
Release Notes: With failure_mode_allowed enabled and external grpc authz server returning error codes like internal error, the request will be allowed (fail-open), see https://github.com/envoyproxy/envoy/issues/34705 for details.
Platform Specific Features: ---
Optional Runtime guard: I'm not sure if it should be guarded. It affects end users, but it's a bug fix.
Fixes #34705

htuch commented:
Probably needs at least runtime guarding given potential for differences in dataplane behavior. @tyxia should the semantics here be harmonized with ext_proc, in particular if we start using it for auth purposes (CC @rshriram)?

ggreenway commented on source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:
```diff
@@ -114,7 +114,9 @@ void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::service::auth::v3::CheckRe
       copyOkResponseMutations(authz_response, ok_response);
     }
-  } else {
+  } else if (response->status().code() == Grpc::Status::WellKnownGrpcStatus::PermissionDenied ||
```
Please add a runtime guard for this behavior change.

konstantin-baidin-y42 replied:
I added a runtime guard and a changelog entry. Can you please take a look?

ggreenway reviewed (changes requested):
Thanks for working on this!

This change needs a runtime guard around the behavior change, and a changelog entry.

/wait

ggreenway commented on test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc:
```diff
@@ -436,6 +437,46 @@ TEST_F(ExtAuthzGrpcClientTest, AuthorizationOkWithAppendActions) {
+TEST_F(ExtAuthzGrpcClientTest, AuthorizationReturnsErrorOnBadGrpcCode) {
+  TestScopedRuntime scoped_runtime;
+  scoped_runtime.mergeValues(
+      {{"envoy.reloadable_features.process_ext_authz_grpc_error_codes_as_errors", "true"}});
```
Please add a second test case that verifies this fails when the runtime guard is set to false.

ggreenway commented on changelogs/current.yaml:
```diff
@@ -246,6 +246,10 @@ bug_fixes:
     Bumped the version of datadog to resolve a crashing bug in earlier versions of the library.
+- area: ext_authz
+  change: |
+    Fixed fail-open behaviour of the failure_mode_allow config option when a grpc external authz server is used.
```
Please add a link to the `failure_mode_allow` field in the docs

konstantin-baidin-y42 replied:
Looks like the https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto page doesn't have anchor links, so I added just a link to the page where this param can be found

ggreenway reviewed (changes requested):
Please clarify in the docs for `failure_mode_allow` (in ext_authz.proto) which grpc response codes are still treated as `denied` when this is true.

/wait

konstantin-baidin-y42 commented:
Hello @ggreenway, thank you for the review.

- I clarified in the docs for failure_mode_allow (in ext_authz.proto) which grpc response codes are still treated as denied when this is true. 
- The runtime guard is now true by default.
- I replaced the old test `AuthorizationDeniedGrpcUnknownStatus` with the new test `AuthorizationReturnsDeniedOnUnknownGrpcCodeWhenFeatureFlagDisabled`
- I tried to add a link for failure_mode_allow docs, but it looks like the https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto page doesn't have anchor links, so I added just a link to the page where this param can be found.

Can you please review again?

htuch reviewed (commented):
/lgtm api

ggreenway commented on changelogs/current.yaml:
```diff
@@ -267,6 +267,12 @@ bug_fixes:
     Fixed a bug where the user data will reference a dangling pointer to the Lua state and cause a crash.
+- area: ext_authz
+  change: |
+    Fixed fail-open behaviour of the failure_mode_allow config option
+    https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto
```
Sorry I wasn't clear about the link. It should be something like 
```suggestion
    Fixed fail-open behaviour of the :ref:`failure_mode_allow config option 
    <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.ExtAuthz.failure_mode_allow>`
```

ggreenway reviewed (changes requested):
/wait

ggreenway reviewed (dismissed):
LGTM, but has a conflict on the changelog. Please resolve, then it's ready to merge.

/wait

konstantin-baidin-y42 commented:
Resolved conflict
